File: src/main/java/at/chaoticbits/config/DecimalFormatter.java
package at.chaoticbits.config;

import java.math.BigDecimal;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;


/**
 * Decimal Format Helper Class.
 * Uses '.' for decimal separator and ',' for grouping thousands
 */
public class DecimalFormatter {


    private static DecimalFormatSymbols decimalFormatSymbols = new DecimalFormatSymbols();
    private static volatile DecimalFormatter instance;

    private DecimalFormatter() {
        decimalFormatSymbols.setDecimalSeparator('.');
        decimalFormatSymbols.setGroupingSeparator(',');
    }

    /**
     * Singleton
     *
     * @return Return the instance of this class
     */
    public static DecimalFormatter getInstance() {
        DecimalFormatter currentInstance;

        if (instance == null) {

            synchronized (DecimalFormatter.class) {
                if (instance == null)
                    instance = new DecimalFormatter();

                currentInstance = instance;
            }

        } else
            currentInstance = instance;

        return currentInstance;
    }

    /**
     * Format price according to decimal dimension in USD
     * @param price price as BigDecimal
     * @return formatted string price
     */
    public static String formatPrice(BigDecimal price) {
        return formatPrice(price, '$');
    }


    /**
     * Format price according to decimal dimension in the given symbol
     * @param price price as BigDecimal
     * @param symbol currency symbol (EUR, USD,..)
     * @return formatted string price
     */
    public static String formatPrice(BigDecimal price, Character symbol) {

        if(price == null)
            return "-";

        DecimalFormat df;

        if (price.multiply(new BigDecimal("1")).compareTo(BigDecimal.ONE) > 0)
            df = new DecimalFormat("#,###.00");
        else
            df = new DecimalFormat("0.00000000");

        df.setDecimalFormatSymbols(decimalFormatSymbols);

        return " " + symbol + df.format(price);
    }


    /**
     * Format percentages according to decimal dimension
     * @param percentage percentage as BigDecimal
     * @return formatted string percentages
     */
    public static String formatPercentage(BigDecimal percentage) {
        if(percentage == null)
            return "-";

        DecimalFormat df;

        if (percentage.multiply(new BigDecimal("1")).compareTo(BigDecimal.ONE) > 0)
            df = new DecimalFormat("#,###.00");
        else
            df = new DecimalFormat("0.00");

        df.setDecimalFormatSymbols(decimalFormatSymbols);

        return  " " + df.format(percentage) + "%";
    }
}

File: src/main/kotlin/at/chaoticbits/config/DecimalFormatter.kt
package at.chaoticbits.config

import java.math.BigDecimal
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols


/**
 * Decimal Format Helper Class.
 * Uses '.' for decimal separator and ',' for grouping thousands
 */
object DecimalFormatter {


    private val decimalFormatSymbols = DecimalFormatSymbols()

    init {
        decimalFormatSymbols.decimalSeparator = '.'
        decimalFormatSymbols.groupingSeparator = ','
    }


    /**
     * Format price according to decimal dimension in USD
     *
     * @param price [BigDecimal] Price as BigDecimal
     * @return [String] Formatted price
     */
    fun formatPrice(price: BigDecimal?): String {
        return if (price == null) "-" else formatPrice(price, '$')
    }


    /**
     * Format price according to decimal dimension in the given symbol
     *
     * @param price [BigDecimal] Price as BigDecimal
     * @param symbol [Char] Currency symbol (EUR, USD,..)
     * @return [String] Formatted price
     */
    fun formatPrice(price: BigDecimal?, symbol: Char = '$'): String {

        if (price == null)
            return "-"

        val df: DecimalFormat = when {
            price.multiply(BigDecimal("1")) > BigDecimal.ONE -> DecimalFormat("#,###.00")
            else -> DecimalFormat("0.00000000")
        }

        df.decimalFormatSymbols = decimalFormatSymbols

        return  " $symbol${df.format(price)}"
    }


    /**
     * Format percentages according to decimal dimension
     *
     * @param percentage [BigDecimal] Percentage
     * @return [String] Formatted percentages
     */
    fun formatPercentage(percentage: BigDecimal?): String {


        if (percentage == null)
            return "-"

        val df: DecimalFormat = when {
            percentage.multiply(BigDecimal("1")) > BigDecimal.ONE -> DecimalFormat("#,###.00")
            else -> DecimalFormat("0.00")
        }

        df.decimalFormatSymbols = decimalFormatSymbols

        return " ${df.format(percentage)}%"

    }
}
